library/core/src/array/mod.rs

```diff
@@ -916,6 +898,24 @@ where
     Ok(Try::from_output(output))
 }
 
+pub(crate) struct Guard<'a, T, const N: usize> {
+    pub array_mut: &'a mut [MaybeUninit<T>; N],
+    pub initialized: usize,
+}
+
+impl<T, const N: usize> Drop for Guard<'_, T, N> {
+    fn drop(&mut self) {
+        debug_assert!(self.initialized <= N);
+
+        // SAFETY: this slice will contain only initialized objects.
```

Comment:
Seems like this is a pretty lightly upheld condition, given the initialized is public (to the whole crate). Not very happy with that. But I'm not sure we have good options here...

Reply:
I could add a bunch of unsafe methods instead. If only we had `pub(unsafe)` for field accesses...

Reply:
I guess this seems like an ArrayVec abstraction anyway, just without all the APIs associated? Do we have traction on that (or perhaps have an impl there already)?

Reply:
I'd be happier with an increment() unsafe method which seems like the only operation we need in the new code at least.

Reply:
> Do we have traction on that (or perhaps have an impl there already)?

There's [a draft RFC](https://github.com/rust-lang/rfcs/pull/3316) but I think that's stuck in the "yes it would be great to have one, but before we do that there's this other allocator topic to research" hell.

We do have some building blocks of an arrayvec floating around, but not the whole thing. And one of the building blocks (`try_collect_into_array`) happens to not optimize well enough in some cases which is why I ended up handrolling it in more direct, unsafe ways in several PRs.

> I'd be happier with an increment() unsafe method which seems like the only operation

In principle touching any of the fields is unsafe, but ok, as long as we have something in the caller telling us that unsafe stuff happens that might be good enough.

Reply:
Okay. It seems like internal ArrayVec shouldn't be controversial even if we don't expose it, but the optimizer point doesn't surprise me (though is also unfortunate).

r=me with increment added, I think it's fine for the time being, even though it's of course true that access to the array also ought to be unsafe.

Reply:
Added documentation and an `unsafe fn push_unchecked()` that does more than just increment.